File: daiphat-fe/src/client/services/supportTicketService.ts
import { apiApp } from '../../api';
import { ApiResponse, PageResponse } from '../../types/api.type';
import {
    CreateSupportTicketCommentRequest,
    CreateSupportTicketRequest,
    GetMyTicketsParams,
    OrderComplaintEligibilityResponse,
    SupportTicketCommentResponse,
    SupportTicketResponse,
    SupportTicketSummaryResponse,
    TicketCategoryResponse,
    UpdateSupportTicketRequest,
} from '../../types/support.type';

const BASE_URL = '/tickets';
const CATEGORY_URL = '/ticket-categories';

const buildMultipartBody = (data: CreateSupportTicketRequest | UpdateSupportTicketRequest, file?: File | null) => {
    const formData = new FormData();
    formData.append('data', new Blob([JSON.stringify(data)], { type: 'application/json' }));
    if (file) {
        formData.append('file', file);
    }
    return formData;
};

const multipartHeaders = {
    headers: { 'Content-Type': 'multipart/form-data' },
};

export const supportTicketService = {
    getCategories: async (): Promise<ApiResponse<TicketCategoryResponse[]>> => {
        const response = await apiApp.get(CATEGORY_URL);
        return response.data;
    },

    getOrderComplaintEligibility: async (
        orderId: string
    ): Promise<ApiResponse<OrderComplaintEligibilityResponse>> => {
        const response = await apiApp.get(`${BASE_URL}/orders/${orderId}/complaint-eligibility`);
        return response.data;
    },

    create: async (
        data: CreateSupportTicketRequest,
        file?: File | null
    ): Promise<ApiResponse<SupportTicketResponse>> => {
        const response = await apiApp.post(
            BASE_URL,
            buildMultipartBody(data, file),
            multipartHeaders
        );
        return response.data;
    },

    getMyTickets: async (
        params: GetMyTicketsParams
    ): Promise<ApiResponse<PageResponse<SupportTicketSummaryResponse>>> => {
        const response = await apiApp.get(`${BASE_URL}/my`, { params });
        return response.data;
    },

    getMyActiveCount: async (): Promise<ApiResponse<number>> => {
        const response = await apiApp.get(`${BASE_URL}/my/active-count`);
        return response.data;
    },

    getById: async (id: number): Promise<ApiResponse<SupportTicketResponse>> => {
        const response = await apiApp.get(`${BASE_URL}/${id}`, {
            skipGlobalErrorToast: true,
        } as any);
        return response.data;
    },

    update: async (
        id: number,
        data: UpdateSupportTicketRequest,
        file?: File | null
    ): Promise<ApiResponse<SupportTicketResponse>> => {
        const response = await apiApp.patch(
            `${BASE_URL}/${id}`,
            buildMultipartBody(data, file),
            multipartHeaders
        );
        return response.data;
    },

    close: async (id: number): Promise<ApiResponse<SupportTicketResponse>> => {
        const response = await apiApp.patch(`${BASE_URL}/${id}/close`);
        return response.data;
    },

    submitResolutionFeedback: async (
        id: number,
        satisfied: boolean
    ): Promise<ApiResponse<SupportTicketResponse>> => {
        const response = await apiApp.put(`${BASE_URL}/${id}/resolution-feedback`, { satisfied });
        return response.data;
    },

    getComments: async (id: number): Promise<ApiResponse<SupportTicketCommentResponse[]>> => {
        const response = await apiApp.get(`${BASE_URL}/${id}/comments`);
        return response.data;
    },

    addComment: async (
        id: number,
        data: CreateSupportTicketCommentRequest,
        file?: File | null
    ): Promise<ApiResponse<SupportTicketCommentResponse>> => {
        const formData = new FormData();
        formData.append('data', new Blob([JSON.stringify(data)], { type: 'application/json' }));
        if (file) {
            formData.append('file', file);
        }
        const response = await apiApp.post(`${BASE_URL}/${id}/comments`, formData, multipartHeaders);
        return response.data;
    },
};

File: daiphat_mobile/lib/src/features/profile/data/support_ticket_service.dart
import 'dart:convert';

import 'package:dio/dio.dart';

import 'package:daiphat_mobile/src/features/profile/data/models/support_ticket.dart';
import 'package:daiphat_mobile/src/shared/network/api_client.dart';
import 'package:daiphat_mobile/src/shared/network/api_exception.dart';

/// Payload tạo/cập nhật một khiếu nại.
class SupportTicketFormData {
  final int ticketCategoryId;
  final String title;
  final String description;
  final String? refId;
  final String? refType;

  const SupportTicketFormData({
    required this.ticketCategoryId,
    required this.title,
    required this.description,
    this.refId,
    this.refType,
  });

  Map<String, dynamic> toJson() => {
        'ticketCategoryId': ticketCategoryId,
        'title': title,
        'description': description,
        if (refId != null && refId!.isNotEmpty) 'refId': refId,
        if (refType != null && refType!.isNotEmpty) 'refType': refType,
      };
}

class SupportTicketService {
  final ApiClient _apiClient;

  SupportTicketService(this._apiClient);

  Future<List<TicketCategoryResponse>> getCategories() async {
    final response = await _apiClient.get('/ticket-categories');
    final data = response['data'] as List<dynamic>? ?? const [];
    return data
        .map((e) => TicketCategoryResponse.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<SupportTicketPageResult> getMyTickets({
    int page = 1,
    int limit = 10,
    String? status,
    String? search,
  }) async {
    final response = await _apiClient.get(
      '/tickets/my',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (status != null && status.isNotEmpty) 'status': status,
        if (search != null && search.isNotEmpty) 'search': search,
      },
    );
    final data = response['data'];
    if (data is! Map<String, dynamic>) {
      throw ApiException(
        response['message']?.toString().isNotEmpty == true
            ? response['message'].toString()
            : 'Không thể tải danh sách khiếu nại.',
      );
    }
    return SupportTicketPageResult.fromJson(data);
  }

  Future<int> getMyActiveCount() async {
    try {
      final response = await _apiClient.get('/tickets/my/active-count');
      final data = response['data'];
      if (data is int) return data;
      return int.tryParse(data?.toString() ?? '') ?? 0;
    } catch (_) {
      return 0;
    }
  }

  Future<SupportTicketResponse> getById(int id) async {
    final response = await _apiClient.get('/tickets/$id');
    final data = response['data'];
    if (data is! Map<String, dynamic>) {
      throw ApiException(
        response['message']?.toString().isNotEmpty == true
            ? response['message'].toString()
            : 'Không tìm thấy khiếu nại.',
      );
    }
    return SupportTicketResponse.fromJson(data);
  }

  Future<SupportTicketResponse> create(
    SupportTicketFormData data, {
    String? filePath,
  }) async {
    final formData = await _buildMultipart(data.toJson(), filePath);
    final response = await _apiClient.post('/tickets', data: formData);
    final result = response['data'];
    if (result is! Map<String, dynamic>) {
      throw ApiException(
        response['message']?.toString().isNotEmpty == true
            ? response['message'].toString()
            : 'Không thể tạo khiếu nại.',
      );
    }
    return SupportTicketResponse.fromJson(result);
  }

  Future<SupportTicketResponse> update(
    int id,
    SupportTicketFormData data, {
    String? filePath,
  }) async {
    final formData = await _buildMultipart(data.toJson(), filePath);
    final response = await _apiClient.patch('/tickets/$id', data: formData);
    final result = response['data'];
    if (result is! Map<String, dynamic>) {
      throw ApiException(
        response['message']?.toString().isNotEmpty == true
            ? response['message'].toString()
            : 'Không thể cập nhật khiếu nại.',
      );
    }
    return SupportTicketResponse.fromJson(result);
  }

  Future<SupportTicketResponse> close(int id) async {
    final response = await _apiClient.patch('/tickets/$id/close');
    final data = response['data'];
    if (data is! Map<String, dynamic>) {
      throw ApiException(
        response['message']?.toString().isNotEmpty == true
            ? response['message'].toString()
            : 'Không thể huỷ khiếu nại.',
      );
    }
    return SupportTicketResponse.fromJson(data);
  }

  Future<SupportTicketResponse> submitResolutionFeedback(
    int id,
    bool satisfied,
  ) async {
    final response = await _apiClient.put(
      '/tickets/$id/resolution-feedback',
      data: {'satisfied': satisfied},
    );
    final data = response['data'];
    if (data is! Map<String, dynamic>) {
      throw ApiException(
        response['message']?.toString().isNotEmpty == true
            ? response['message'].toString()
            : 'Không thể gửi phản hồi.',
      );
    }
    return SupportTicketResponse.fromJson(data);
  }

  Future<List<SupportTicketCommentResponse>> getComments(int id) async {
    final response = await _apiClient.get('/tickets/$id/comments');
    final data = response['data'] as List<dynamic>? ?? const [];
    return data
        .map((e) =>
            SupportTicketCommentResponse.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<SupportTicketCommentResponse> addComment(
    int id,
    String content, {
    String? filePath,
  }) async {
    final formData = await _buildMultipart({'content': content}, filePath);
    final response =
        await _apiClient.post('/tickets/$id/comments', data: formData);
    final data = response['data'];
    if (data is! Map<String, dynamic>) {
      throw ApiException(
        response['message']?.toString().isNotEmpty == true
            ? response['message'].toString()
            : 'Không thể gửi bình luận.',
      );
    }
    return SupportTicketCommentResponse.fromJson(data);
  }

  Future<FormData> _buildMultipart(
    Map<String, dynamic> data,
    String? filePath,
  ) async {
    final formData = FormData();
    formData.files.add(
      MapEntry(
        'data',
        MultipartFile.fromString(
          jsonEncode(data),
          contentType: DioMediaType('application', 'json'),
        ),
      ),
    );
    if (filePath != null && filePath.isNotEmpty) {
      formData.files.add(
        MapEntry('file', await MultipartFile.fromFile(filePath)),
      );
    }
    return formData;
  }
}
